graylog2-server/src/main/java/org/graylog2/migrations/V20211221144300_GeoIpResolverConfigMigration.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+
+package org.graylog2.migrations;
+
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Updates;
+import com.mongodb.client.result.UpdateResult;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.graylog.plugins.map.config.DatabaseVendorType;
+import org.graylog.plugins.map.config.GeoIpResolverConfig;
+import org.graylog2.database.MongoConnection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.time.LocalDateTime;
+import java.time.Month;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+
+public class V20211221144300_GeoIpResolverConfigMigration extends Migration {
+
+    private static final Logger LOG = LoggerFactory.getLogger(V20211221144300_GeoIpResolverConfigMigration.class);
+    private static final String COLLECTION_NAME = "cluster_config";
+    public static final String PAYLOAD = "payload";
+    private static final String FIELD_DB_VENDOR = PAYLOAD + ".database_vendor_type";
+    private static final String FIELD_DB_TYPE = PAYLOAD + ".db_type";
+
+    private final MongoConnection mongoConnection;
+
+    @Inject
+    public V20211221144300_GeoIpResolverConfigMigration(MongoConnection mongoConnection) {
+        this.mongoConnection = mongoConnection;
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.of(LocalDateTime.of(2021, Month.DECEMBER, 21, 14, 43), ZoneId.systemDefault());
```

Comment:
> I added a few questions and suggestions below.
> 
> In your [comment](https://github.com/Graylog2/graylog2-server/pull/11854/files#r774233058) you mentioned the management of the database reader.
> 
> How is lifecycle management handled in general for the objects that read from a database file?
> 
> I don't see any code that "restarts" the objects once a user changes the configuration in the web interface. So it looks like any change to the resolver config currently requires a restart of the server. Everything that is configured in the web interface shouldn't require a restart to get applied.
> 
> Lookup tables, for example, are services with lifecycle management that will be restarted once their config gets updated. (that includes closing file handles to databases, etc. - see: [MaxmindDataAdapter](https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog/plugins/map/geoip/MaxmindDataAdapter.java))
> 
> To make the runtime changes work, we would need to implement something similar for the geo IP resolver engine.
> 
> And this raises the question of what use case or problem we need to solve triggered these changes in the first place? sweat_smile If we add lifecycle handling to the geo IP resolver engine, we basically reimplement parts of lookup tables.
> 
> So I think we should take a step back and talk about the problem we want to solve. /cc @waab76 @bud1979

@bernd , thanks for all your feedback.  As far as _restarting the objects_, this is done by the GeoIPProcessor::updateConfig method, which subscribes to a message posted by the ClusterConfigService after it writes the new config data.

I discovered the **MaxMindDataAdapter** yesterday (well, @mikedklein did actually), so I have not looked in detail how it will be affected, nor if changes to that will/should be included in this code change or if it should be in a separate issue.

Thanks again for all your feedback.

Reply:
> As far as _restarting the objects_, this is done by the GeoIPProcessor::updateConfig method, which subscribes to a message posted by the ClusterConfigService after it writes the new config data.

Doh! I even wrote that code. :see_no_evil: :laughing: Thanks for the update!